### PR TITLE
MNIST-C & MVTec-AD

### PR DESCRIPTION
**Pull request by vitskvara**
I will also add MvTec-AD here.

MNIST-C consists of 15 corruption categories that change the original MNIST dataset. This basically presents us with 15 new datasets, where we will detect non-corrupted vs corrupted.

 "brightness"
 "canny_edges"
 "dotted_line"
 "fog"
 "glass_blur"
 "impulse_noise"
 "motion_blur"
 "rotate"
 "scale"
 "shear"
 "shot_noise"
 "spatter"
 "stripe"
 "translate"
 "zigzag"

### CODE REVIEW

**Comment by vitskvara** [comment]
MVTec-AD consists of 15 subproblems. Each has a few (10s~) normal samples. Each subproblem has a set of different kinds of anomalies, e.g. in `bottle`, there are `broken_large,  broken_small,  contamination` kinds of anomalies. The images are large - 900x900x3, but there is not a lot of them.

bottle 
capsule
grid      
leather      
metal_nut 
tile        
transistor  
zipper
cable   
carpet   
hazelnut 
pill       
screw       
toothbrush  
wood

**Comment by vitskvara** [comment]
Currently the else-if in `load_data` is really ugly, but it seems to be the shortest way to do it.

**Comment by vitskvara** [comment]
VAE on MNIST-C seems to be working.

**Comment by vitskvara** [comment]
But I am getting OOM errors for MVTec (on the RCI cluster).

**Comment by vitskvara** [comment]
I am currently using a Dense layer on the output of decoder that comes after the convolutions and which I use to compute the vectorized mean and scalar variance of the output. However, this now tries to use a W matrix with one dimension of 3x900^2, which already fails at initialization with Glorot. So I will rewrite this piece without the last Dense layer, probably estimating the scalar variance from an additional conv channel.

File: scripts/experiments_images/run_parallel_no_class.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# This runs parallel experiments over all datasets.
+# USAGE EXAMPLE
+# 	./run_parallel.sh pidforest 3 1 2 datasets_tabular.txt 0.01
```
Comment (janfrancu) [inline scripts/experiments_images/run_parallel_no_class.sh]: Please update the usage example.

File: src/datasets/basics.jl [hunk 2]
```diff
@@ -93,8 +93,20 @@ function load_data(dataset::String, ratios=(0.6,0.2,0.2); seed=nothing,
...
-	elseif dataset in other_datasets
+	elseif dataset in other_datasets # other tabular datasets
 		data_normal, data_anomalous = load_other_data(dataset; standardize=true, kwargs...)
+	elseif dataset =="MNIST-C"
+		data_normal, data_anomalous = load_mnist_c_data(; kwargs...)
+	elseif dataset == "MVTec-AD"
+		data_normal, data_anomalous = load_mvtec_ad_data(; kwargs...)
+	elseif occursin("MNIST-C", dataset)
+		sd = split(dataset, "_")
+		category = (length(sd) >= 2) ? dataset[9:end] : "brightness"
```
Comment (janfrancu) [inline src/datasets/basics.jl]: Add comment as to why this case has to be handled separately.

Comment (janfrancu) [reply src/datasets/basics.jl]: Correct me if I am wrong, but looking at the possibilities in `scripts/experiments_images/datasets_mnistc.txt` this never chooses `MNIST-C_fog` or others without underscore.

File: src/datasets/basics.jl [hunk 3]
```diff
@@ -93,8 +93,20 @@ function load_data(dataset::String, ratios=(0.6,0.2,0.2); seed=nothing,
...
+		data_normal, data_anomalous = load_mnist_c_data(; category=category, kwargs...)
+	elseif occursin("MVTec-AD", dataset) 
+		sd = split(dataset, "_")
+		category = (length(sd) >= 2) ? dataset[10:end] : "bottle"
```
Comment (janfrancu) [inline src/datasets/basics.jl]: Not sure about this either.

File: src/datasets/img.jl [hunk 4]
```diff
@@ -28,3 +28,79 @@ function load_mldatasets_data(dataset::String; anomaly_class_ind::Int=1)
...
+
+"""
+	load_images_rgb(path)
+
+Load all images in a path.
+"""
+function load_images_rgb(path)
+	imgs = []
+	for f in readdir(path, join=true)
+		try # some files are corrupted
```
Comment (janfrancu) [inline src/datasets/img.jl:98]: Has this already been reported somewhere or are we the first one to encounter it?

**Review by janfrancu — CHANGES_REQUESTED** [review]

File: src/datasets/basics.jl (hunk 2 above)
Comment (vitskvara) [reply src/datasets/basics.jl]:
```julia
dataset = "MNIST-C_fog"
sd = split(dataset, "_")
category = (length(sd) >= 2) ? dataset[9:end] : "brightness"
```
`category` now evaluates to `"fog"`

File: scripts/experiments_images/run_parallel_no_class.sh (hunk 1 above)
Comment (vitskvara) [reply scripts/experiments_images/run_parallel_no_class.sh]: Done

File: src/datasets/img.jl (hunk 4 above)
Comment (vitskvara) [reply src/datasets/img.jl:98]: Ok, so this was a problem with my partial download of the dataset. Should I remove the try catch block?

**Review by janfrancu — APPROVED** [review]

**Pull request merged**